File: test/shire/agents_test.exs
defmodule Shire.AgentsTest do
  use Shire.DataCase, async: true

  import Mox

  alias Shire.Agents
  alias Shire.Projects

  @vm Shire.VirtualMachineStub

  setup :set_mox_from_context

  setup do
    stub(Shire.VirtualMachineMock, :workspace_root, fn _project_id -> "/workspace" end)

    {:ok, project} = Projects.create_project("test-project")
    {:ok, agent} = Agents.create_agent_with_vm(project.id, "test-agent", "version: 1\n", @vm)
    {:ok, agent2} = Agents.create_agent_with_vm(project.id, "chat-agent", "version: 1\n", @vm)

    %{project: project, agent: agent, agent2: agent2}
  end

  describe "messages" do
    test "create_message/1 with valid data", %{project: project, agent: agent} do
      assert {:ok, msg} =
               Agents.create_message(%{
                 project_id: project.id,
                 agent_id: agent.id,
                 role: "user",
                 content: %{"text" => "hi"}
               })

      assert msg.role == "user"
      assert msg.content == %{"text" => "hi"}
      assert msg.agent_id == agent.id
      assert msg.project_id == project.id
    end

    test "create_message/1 requires project_id, agent_id and role" do
      assert {:error, changeset} = Agents.create_message(%{})
      assert "can't be blank" in errors_on(changeset).project_id
      assert "can't be blank" in errors_on(changeset).agent_id
      assert "can't be blank" in errors_on(changeset).role
    end

    test "list_messages_for_agent/2 returns messages oldest first", %{
      project: project,
      agent2: agent
    } do
      {:ok, _} =
        Agents.create_message(%{
          project_id: project.id,
          agent_id: agent.id,
          role: "user",
          content: %{"text" => "first"}
        })

      {:ok, _} =
        Agents.create_message(%{
          project_id: project.id,
          agent_id: agent.id,
          role: "agent",
          content: %{"text" => "second"}
        })

      {messages, _has_more} = Agents.list_messages_for_agent(project.id, agent.id)
      assert length(messages) == 2
      assert Enum.at(messages, 0).content["text"] == "first"
      assert Enum.at(messages, 1).content["text"] == "second"
    end

    test "list_inter_agent_messages/1 returns only inter_agent messages", %{
      project: project,
      agent2: agent
    } do
      {:ok, _} =
        Agents.create_message(%{
          project_id: project.id,
          agent_id: agent.id,
          role: "user",
          content: %{"text" => "hi"}
        })

      {:ok, _} =
        Agents.create_message(%{
          project_id: project.id,
          agent_id: agent.id,
          role: "inter_agent",
          content: %{
            "text" => "Hello from Alice",
            "from_agent" => "Alice",
            "to_agent" => "chat-agent"
          }
        })

      {messages, _has_more} = Agents.list_inter_agent_messages(project.id)
      assert length(messages) == 1
      assert hd(messages).role == "inter_agent"
      assert hd(messages).content["from_agent"] == "Alice"
    end

    test "list_inter_agent_messages/2 supports cursor-based pagination", %{
      project: project,
      agent2: agent
    } do
      for i <- 1..3 do
        Agents.create_message(%{
          project_id: project.id,
          agent_id: agent.id,
          role: "inter_agent",
          content: %{
            "text" => "msg-#{i}",
            "from_agent" => "Alice",
            "to_agent" => "chat-agent"
          }
        })
      end

      {first_page, has_more} = Agents.list_inter_agent_messages(project.id, limit: 2)
      assert length(first_page) == 2
      assert has_more

      oldest_id = List.last(first_page).id

      {second_page, has_more2} =
        Agents.list_inter_agent_messages(project.id, before: oldest_id, limit: 2)

      assert length(second_page) == 1
      refute has_more2
    end

    test "send_message_with_inbox/6 inserts message and writes inbox file", %{
      project: project,
      agent: agent
    } do
      inbox_path = "/workspace/agents/#{agent.id}/inbox/msg-1.yaml"
      envelope = %{"role" => "user", "content" => "hello"}

      assert {:ok, msg} =
               Agents.send_message_with_inbox(
                 project.id,
                 agent.id,
                 "hello",
                 inbox_path,
                 envelope,
                 @vm
               )

      assert msg.role == "user"
      assert msg.content == %{"text" => "hello"}
      assert msg.agent_id == agent.id
      assert msg.project_id == project.id
    end

    test "send_message_with_inbox/6 rolls back message on VM write failure", %{
      project: project,
      agent: agent
    } do
      defmodule FailingWriteVM do
        @behaviour Shire.VirtualMachine
        def cmd(_p, _c, _a \\ [], _o \\ []), do: {:ok, ""}
        def cmd!(_p, _c, _a \\ [], _o \\ []), do: ""
        def read(_p, _path), do: {:ok, ""}
        def write(_p, _path, _content), do: {:error, :disk_full}
        def mkdir_p(_p, _path), do: :ok
        def rm(_p, _path), do: :ok
        def rm_rf(_p, _path), do: :ok
        def ls(_p, _path), do: {:ok, []}
        def stat(_p, _path), do: {:ok, %{type: :file, size: 0}}
        def spawn_command(_p, _c, _a \\ [], _o \\ []), do: {:error, :not_available}
        def write_stdin(_c, _d), do: :ok
        def resize(_c, _r, _cols), do: :ok

        def workspace_root(_p), do: "/workspace"
        def touch_keepalive(_p), do: :ok
        def vm_status(_p), do: :running
        def destroy_vm(_p), do: :ok
      end

      inbox_path = "/workspace/agents/#{agent.id}/inbox/msg-2.yaml"
      envelope = %{"role" => "user", "content" => "fail"}

      assert {:error, :disk_full} =
               Agents.send_message_with_inbox(
                 project.id,
                 agent.id,
                 "fail",
                 inbox_path,
                 envelope,
                 FailingWriteVM
               )

      # Message should not have been persisted
      {messages, _} = Agents.list_messages_for_agent(project.id, agent.id)
      refute Enum.any?(messages, &(&1.content["text"] == "fail"))
    end

    test "send_message_with_inbox stores attachments in content map", %{
      project: project,
      agent: agent
    } do
      inbox_path = "/workspace/agents/#{agent.id}/inbox/msg-att.yaml"
      envelope = %{"role" => "user", "content" => "with files"}

      attachments = [
        %{
          "id" => "abc12345",
          "filename" => "report.pdf",
          "size" => 1024,
          "content_type" => "application/pdf"
        }
      ]

      assert {:ok, msg} =
               Agents.send_message_with_inbox(
                 project.id,
                 agent.id,
                 "check this file",
                 inbox_path,
                 envelope,
                 @vm,
                 attachments: attachments
               )

      assert msg.content["text"] == "check this file"
      assert msg.content["attachments"] == attachments
    end

    test "send_message_with_inbox omits attachments key when empty", %{
      project: project,
      agent: agent
    } do
      inbox_path = "/workspace/agents/#{agent.id}/inbox/msg-noatt.yaml"
      envelope = %{"role" => "user", "content" => "no files"}

      assert {:ok, msg} =
               Agents.send_message_with_inbox(
                 project.id,
                 agent.id,
                 "just text",
                 inbox_path,
                 envelope,
                 @vm,
                 attachments: []
               )

      assert msg.content == %{"text" => "just text"}
      refute Map.has_key?(msg.content, "attachments")
    end

    test "unread_counts/2 returns 0 for agents with no messages", %{
      agent: agent,
      agent2: agent2
    } do
      agents = [
        %{id: agent.id, last_read_message_id: nil},
        %{id: agent2.id, last_read_message_id: nil}
      ]

      counts = Agents.unread_counts(agents)
      assert Map.get(counts, agent.id) == 0
      assert Map.get(counts, agent2.id) == 0
    end

    test "unread_counts/2 counts only agent role messages", %{
      project: project,
      agent: agent
    } do
      # User message — should not count
      {:ok, _} =
        Agents.create_message(%{
          project_id: project.id,
          agent_id: agent.id,
          role: "user",
          content: %{"text" => "hello"}
        })

      # Tool use — should not count
      {:ok, _} =
        Agents.create_message(%{
          project_id: project.id,
          agent_id: agent.id,
          role: "tool_use",
          content: %{"tool" => "read", "tool_use_id" => "t1"}
        })

      # Inter-agent message — should not count
      {:ok, _} =
        Agents.create_message(%{
          project_id: project.id,
          agent_id: agent.id,
          role: "inter_agent",
          content: %{"text" => "peer msg", "from_agent" => "other"}
        })

      # Agent message — should count
      {:ok, _} =
        Agents.create_message(%{
          project_id: project.id,
          agent_id: agent.id,
          role: "agent",
          content: %{"text" => "response"}
        })

      agents = [%{id: agent.id, last_read_message_id: nil}]
      counts = Agents.unread_counts(agents)
      assert Map.get(counts, agent.id) == 1
    end

    test "unread_counts/2 returns 0 when last_read is at latest message", %{
      project: project,
      agent: agent
    } do
      {:ok, m1} =
        Agents.create_message(%{
          project_id: project.id,
          agent_id: agent.id,
          role: "agent",
          content: %{"text" => "first"}
        })

      {:ok, m2} =
        Agents.create_message(%{
          project_id: project.id,
          agent_id: agent.id,
          role: "agent",
          content: %{"text" => "second"}
        })

      # All read
      agents = [%{id: agent.id, last_read_message_id: m2.id}]
      counts = Agents.unread_counts(agents)
      assert Map.get(counts, agent.id) == 0

      # Only first read
      agents = [%{id: agent.id, last_read_message_id: m1.id}]
      counts = Agents.unread_counts(agents)
      assert Map.get(counts, agent.id) == 1
    end

    test "latest_message_id/1 returns nil when no messages exist", %{agent: agent} do
      assert Agents.latest_message_id(agent.id) == nil
    end

    test "latest_message_id/1 returns the latest agent-role message id", %{
      project: project,
      agent: agent
    } do
      {:ok, _} =
        Agents.create_message(%{
          project_id: project.id,
          agent_id: agent.id,
          role: "user",
          content: %{"text" => "hello"}
        })

      {:ok, m1} =
        Agents.create_message(%{
          project_id: project.id,
          agent_id: agent.id,
          role: "agent",
          content: %{"text" => "first"}
        })

      assert Agents.latest_message_id(agent.id) == m1.id

      {:ok, m2} =
        Agents.create_message(%{
          project_id: project.id,
          agent_id: agent.id,
          role: "agent",
          content: %{"text" => "second"}
        })

      assert Agents.latest_message_id(agent.id) == m2.id
    end

    test "delete_agent_with_vm/3 deletes agent and its messages", %{project: project} do
      {:ok, del_agent} =
        Agents.create_agent_with_vm(project.id, "delete-test", "version: 1\n", @vm)

      {:ok, _} =
        Agents.create_message(%{
          project_id: project.id,
          agent_id: del_agent.id,
          role: "user",
          content: %{"text" => "hi"}
        })

      {:ok, other_agent} =
        Agents.create_agent_with_vm(project.id, "other-agent", "version: 1\n", @vm)

      {:ok, _} =
        Agents.create_message(%{
          project_id: project.id,
          agent_id: other_agent.id,
          role: "user",
          content: %{"text" => "keep me"}
        })

      :ok = Agents.delete_agent_with_vm(project.id, del_agent, @vm)

      assert {:error, :not_found} = Agents.get_agent(del_agent.id)

      {other_msgs, _} = Agents.list_messages_for_agent(project.id, other_agent.id)
      assert length(other_msgs) == 1
    end
  end
end

File: src/services/agents.test.ts
import { describe, it, expect, beforeEach } from "bun:test";
import { useTestDb } from "../test/setup";
import { getDb } from "../db";
import * as agents from "./agents";
import * as projects from "./projects";

describe("agents service", () => {
  useTestDb();

  let projectId: string;
  let agentId: string;
  let agent2Id: string;

  beforeEach(() => {
    const project = projects.createProject("test-project");
    projectId = project.id;
    const agent = agents.createAgent(projectId, { name: "test-agent" });
    agentId = agent.id;
    const agent2 = agents.createAgent(projectId, { name: "chat-agent" });
    agent2Id = agent2.id;
  });

  describe("createMessage", () => {
    it("creates a message with valid data", () => {
      const msg = agents.createMessage({
        projectId,
        agentId,
        role: "user",
        content: { text: "hi" },
      });
      expect(msg.role).toBe("user");
      expect((msg.content as Record<string, unknown>).text).toBe("hi");
      expect(msg.agentId).toBe(agentId);
      expect(msg.projectId).toBe(projectId);
    });
  });

  describe("listMessages", () => {
    it("returns messages oldest first", () => {
      agents.createMessage({
        projectId,
        agentId: agent2Id,
        role: "user",
        content: { text: "first" },
      });
      agents.createMessage({
        projectId,
        agentId: agent2Id,
        role: "agent",
        content: { text: "second" },
      });

      const { messages } = agents.listMessages(projectId, agent2Id);
      expect(messages.length).toBe(2);
      expect((messages[0].content as Record<string, unknown>).text).toBe("first");
      expect((messages[1].content as Record<string, unknown>).text).toBe("second");
    });
  });

  describe("listInterAgentMessages", () => {
    it("returns only inter_agent messages", () => {
      agents.createMessage({
        projectId,
        agentId: agent2Id,
        role: "user",
        content: { text: "hi" },
      });
      agents.createMessage({
        projectId,
        agentId: agent2Id,
        role: "inter_agent",
        content: { text: "Hello from Alice", from_agent: "Alice", to_agent: "chat-agent" },
      });

      const { messages } = agents.listInterAgentMessages(projectId);
      expect(messages.length).toBe(1);
      expect(messages[0].role).toBe("inter_agent");
      expect((messages[0].content as Record<string, unknown>).from_agent).toBe("Alice");
    });

    it("supports cursor-based pagination", () => {
      for (let i = 1; i <= 3; i++) {
        agents.createMessage({
          projectId,
          agentId: agent2Id,
          role: "inter_agent",
          content: { text: `msg-${i}`, from_agent: "Alice", to_agent: "chat-agent" },
        });
      }

      const { messages: firstPage, hasMore } = agents.listInterAgentMessages(projectId, {
        limit: 2,
      });
      expect(firstPage.length).toBe(2);
      expect(hasMore).toBe(true);

      const oldestId = firstPage[firstPage.length - 1].id;
      const { messages: secondPage, hasMore: hasMore2 } = agents.listInterAgentMessages(projectId, {
        before: oldestId,
        limit: 2,
      });
      expect(secondPage.length).toBe(1);
      expect(hasMore2).toBe(false);
    });
  });

  describe("unreadCounts", () => {
    it("runs all queries in a single transaction for snapshot consistency", () => {
      // Create messages for both agents
      agents.createMessage({ projectId, agentId, role: "agent", content: { text: "a1" } });
      agents.createMessage({
        projectId,
        agentId: agent2Id,
        role: "agent",
        content: { text: "b1" },
      });

      const counts = agents.unreadCounts(
        [agentId, agent2Id],
        new Map([
          [agentId, null],
          [agent2Id, null],
        ]),
      );

      // Both agents should have consistent counts from the same snapshot
      expect(counts.get(agentId)).toBe(1);
      expect(counts.get(agent2Id)).toBe(1);
    });

    it("returns 0 for agents with no messages", () => {
      const counts = agents.unreadCounts(
        [agentId, agent2Id],
        new Map([
          [agentId, null],
          [agent2Id, null],
        ]),
      );
      expect(counts.get(agentId)).toBe(0);
      expect(counts.get(agent2Id)).toBe(0);
    });

    it("counts only agent role messages", () => {
      // User message — should not count
      agents.createMessage({ projectId, agentId, role: "user", content: { text: "hello" } });
      // Tool use — should not count
      agents.createMessage({
        projectId,
        agentId,
        role: "tool_use",
        content: { tool: "read", tool_use_id: "t1" },
      });
      // Inter-agent — should not count
      agents.createMessage({
        projectId,
        agentId,
        role: "inter_agent",
        content: { text: "peer msg", from_agent: "other" },
      });
      // Agent message — should count
      agents.createMessage({ projectId, agentId, role: "agent", content: { text: "response" } });

      const counts = agents.unreadCounts([agentId], new Map([[agentId, null]]));
      expect(counts.get(agentId)).toBe(1);
    });

    it("returns 0 when last_read is at latest message", () => {
      const m1 = agents.createMessage({
        projectId,
        agentId,
        role: "agent",
        content: { text: "first" },
      });
      const m2 = agents.createMessage({
        projectId,
        agentId,
        role: "agent",
        content: { text: "second" },
      });

      // All read
      let counts = agents.unreadCounts([agentId], new Map([[agentId, m2.id]]));
      expect(counts.get(agentId)).toBe(0);

      // Only first read
      counts = agents.unreadCounts([agentId], new Map([[agentId, m1.id]]));
      expect(counts.get(agentId)).toBe(1);
    });
  });

  describe("latestAgentMessageId", () => {
    it("returns null when no messages exist", () => {
      expect(agents.latestAgentMessageId(agentId)).toBeNull();
    });

    it("returns the latest agent-role message id", () => {
      agents.createMessage({ projectId, agentId, role: "user", content: { text: "hello" } });
      const m1 = agents.createMessage({
        projectId,
        agentId,
        role: "agent",
        content: { text: "first" },
      });
      expect(agents.latestAgentMessageId(agentId)).toBe(m1.id);

      const m2 = agents.createMessage({
        projectId,
        agentId,
        role: "agent",
        content: { text: "second" },
      });
      expect(agents.latestAgentMessageId(agentId)).toBe(m2.id);
    });
  });

  describe("latestUserMessageAt", () => {
    it("returns null when no user messages exist", () => {
      expect(agents.latestUserMessageAt(agentId)).toBeNull();
    });

    it("returns null when only non-user messages exist", () => {
      agents.createMessage({ projectId, agentId, role: "agent", content: { text: "response" } });
      agents.createMessage({
        projectId,
        agentId,
        role: "inter_agent",
        content: { text: "peer msg", from_agent: "other" },
      });
      expect(agents.latestUserMessageAt(agentId)).toBeNull();
    });

    it("returns the timestamp of the latest user message", () => {
      const m1 = agents.createMessage({
        projectId,
        agentId,
        role: "user",
        content: { text: "first" },
      });
      expect(agents.latestUserMessageAt(agentId)).toBe(m1.createdAt);

      const m2 = agents.createMessage({
        projectId,
        agentId,
        role: "user",
        content: { text: "second" },
      });
      expect(agents.latestUserMessageAt(agentId)).toBe(m2.createdAt);
    });

    it("ignores messages from other agents", () => {
      agents.createMessage({
        projectId,
        agentId: agent2Id,
        role: "user",
        content: { text: "hello" },
      });
      expect(agents.latestUserMessageAt(agentId)).toBeNull();
    });
  });

  describe("getMessage / updateMessage", () => {
    it("gets and updates a message", () => {
      const msg = agents.createMessage({
        projectId,
        agentId,
        role: "tool_use",
        content: { tool: "read", tool_use_id: "t1", output: null },
      });

      const fetched = agents.getMessage(msg.id);
      expect(fetched).toBeTruthy();
      expect((fetched!.content as Record<string, unknown>).tool).toBe("read");

      agents.updateMessage(msg.id, {
        content: { tool: "read", tool_use_id: "t1", output: "done" },
      });
      const updated = agents.getMessage(msg.id);
      expect((updated!.content as Record<string, unknown>).output).toBe("done");
    });
  });

  describe("updateAgent", () => {
    it("updates agent fields", () => {
      const updated = agents.updateAgent(agentId, {
        description: "Updated description",
        model: "claude-sonnet-4-6",
      });
      expect(updated).toBeDefined();
      expect(updated!.description).toBe("Updated description");
      expect(updated!.model).toBe("claude-sonnet-4-6");
    });

    it("updates name without affecting other fields", () => {
      agents.updateAgent(agentId, { description: "Keep me", model: "test-model" });
      const renamed = agents.updateAgent(agentId, { name: "renamed-agent" });
      expect(renamed!.name).toBe("renamed-agent");
      expect(renamed!.description).toBe("Keep me");
      expect(renamed!.model).toBe("test-model");
    });
  });

  describe("setSessionId", () => {
    it("sets sessionId on an agent", () => {
      agents.setSessionId(agentId, "session-abc-123");
      const agent = agents.getAgent(agentId);
      expect(agent!.sessionId).toBe("session-abc-123");
    });

    it("clears sessionId when set to null", () => {
      agents.setSessionId(agentId, "session-abc-123");
      agents.setSessionId(agentId, null);
      const agent = agents.getAgent(agentId);
      expect(agent!.sessionId).toBeNull();
    });
  });

  describe("createAgent with recipe fields", () => {
    it("stores recipe fields in DB", () => {
      const agent = agents.createAgent(projectId, {
        name: "recipe-agent",
        description: "A test agent",
        harness: "claude_code",
        model: "claude-sonnet-4-6",
      });
      expect(agent.description).toBe("A test agent");
      expect(agent.harness).toBe("claude_code");
      expect(agent.model).toBe("claude-sonnet-4-6");
    });
  });

  describe("emoji support", () => {
    it("stores emoji on creation", () => {
      const agent = agents.createAgent(projectId, {
        name: "emoji-agent",
        emoji: "\u{1F680}",
      });
      expect(agent.emoji).toBe("\u{1F680}");
    });

    it("defaults emoji to null when not provided", () => {
      const agent = agents.createAgent(projectId, { name: "no-emoji-agent" });
      expect(agent.emoji).toBeNull();
    });

    it("updates emoji via updateAgent", () => {
      const agent = agents.createAgent(projectId, { name: "update-emoji-agent" });
      const updated = agents.updateAgent(agent.id, { emoji: "\u{1F916}" });
      expect(updated!.emoji).toBe("\u{1F916}");
    });
  });

  describe("transaction support", () => {
    it("createMessage rolls back when transaction fails", () => {
      const initialMessages = agents.listMessages(projectId, agentId).messages.length;

      expect(() => {
        getDb().transaction((tx) => {
          agents.createMessage(
            { projectId, agentId, role: "user", content: { text: "will rollback" } },
            tx,
          );
          throw new Error("simulated failure");
        });
      }).toThrow("simulated failure");

      const afterMessages = agents.listMessages(projectId, agentId).messages.length;
      expect(afterMessages).toBe(initialMessages);
    });

    it("deleteAgent rolls back when transaction fails", () => {
      const tempAgent = agents.createAgent(projectId, { name: "temp-agent" });

      expect(() => {
        getDb().transaction((tx) => {
          agents.deleteAgent(tempAgent.id, tx);
          throw new Error("simulated failure");
        });
      }).toThrow("simulated failure");

      // Agent should still exist
      expect(agents.getAgent(tempAgent.id)).toBeDefined();
    });
  });

  describe("deleteAgent", () => {
    it("deletes agent and cascades messages", () => {
      const delAgent = agents.createAgent(projectId, { name: "delete-test" });
      agents.createMessage({
        projectId,
        agentId: delAgent.id,
        role: "user",
        content: { text: "hi" },
      });

      // Other agent's messages should survive
      const otherAgent = agents.createAgent(projectId, { name: "other-agent" });
      agents.createMessage({
        projectId,
        agentId: otherAgent.id,
        role: "user",
        content: { text: "keep me" },
      });

      agents.deleteAgent(delAgent.id);
      expect(agents.getAgent(delAgent.id)).toBeUndefined();

      const { messages: otherMsgs } = agents.listMessages(projectId, otherAgent.id);
      expect(otherMsgs.length).toBe(1);
    });
  });
});
